src/file/file.rs

```diff
@@ -0,0 +1,282 @@
+        for &indent_prefix in VALID_INDENTS {
+            for &indent_suffix in VALID_INDENTS {
+                let expected = format!("{}{}{}\n", indent_prefix, line, indent_suffix);
+                let actual = parse_line(expected.as_str());
+                assert_eq!(
+                    actual,
+                    Line {
+                        indent_prefix: indent_prefix.to_string(),
+                        expression: Expression::Malformed(line.to_string()),
+                        indent_suffix: indent_suffix.to_string(),
+                    }
+                );
+                assert_eq!(expected, actual.to_string());
+            }
+        }
+    }
+
+    #[test]
+    fn smoke_parse_correct_line_test() {
+        test_correct_line(
+            "Host",
+            "example.com",
+            vec![ArgumentToken::Pure("example.com".to_string())],
+        );
+        test_correct_line("Host", "*", vec![ArgumentToken::Pure("*".to_string())]);
+        test_correct_line(
+            "Host",
+            "lol!*.com example.com \t !kek!",
+            vec![
+                ArgumentToken::Pure("lol!*.com".to_string()),
+                ArgumentToken::Whitespace(" ".to_string()),
+                ArgumentToken::Pure("example.com".to_string()),
+                ArgumentToken::Whitespace(" \t ".to_string()),
+                ArgumentToken::Pure("!kek!".to_string()),
+            ],
+        );
+        test_correct_line(
+            "Host",
+            "\"*\"",
+            vec![ArgumentToken::Quoted("*".to_string())],
+        );
+        test_correct_line(
+            "Host",
+            "\"hello # \\\" lol \"",
+            vec![ArgumentToken::Quoted("hello # \\\" lol ".to_string())],
+        );
+        test_correct_line(
+            "Host",
+            "lol!*.com \"example.com\" \t !kek!",
+            vec![
+                ArgumentToken::Pure("lol!*.com".to_string()),
+                ArgumentToken::Whitespace(" ".to_string()),
+                ArgumentToken::Quoted("example.com".to_string()),
+                ArgumentToken::Whitespace(" \t ".to_string()),
+                ArgumentToken::Pure("!kek!".to_string()),
+            ],
+        );
+    }
+
+    #[test]
+    fn malformed_lines_test() {
+        const INVALID_EXPRESSIONS: &[&str] = &[
+            "Host",
+            "Host0 example.com",
+            "Host # kek",
+            "123",
+            "Ho#st kek",
+            "Host #kek",
+            "Host k#k",
+            "Host kek #",
+            "Host ke\"k # lol",
+            "Host \"lol",
+        ];
+        for &invalid_expression in INVALID_EXPRESSIONS {
+            test_malformed_line(invalid_expression);
+        }
+    }
+
+    #[test]
+    fn deserialization_test() {
+        let config = "# a comment\n\
+                            \t # one more comment \t\r\n\
+                            \t Host example.com \n\
+                            User root\n\
+                            \n"
```

Comment:
let's consider case when file ends not with '\n'. now it seems like idempotency of consequent read and write won't be preserved